napari_btrack/track.py

```diff
@@ -1,24 +1,144 @@
-from pathlib import Path
-from typing import Optional
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Tuple, Union
 
 import btrack
 import napari
+import numpy as np
+import numpy.typing as npt
+from btrack import datasets
+from btrack.config import (
+    HypothesisModel,
+    MotionModel,
+    TrackerConfig,
+    load_config,
+    save_config,
+)
 from btrack.utils import segmentation_to_objects
-from magicgui import magicgui
-from magicgui.widgets import FunctionGui
+from magicgui.application import use_app
+from magicgui.types import FileDialogMode
+from magicgui.widgets import Container, PushButton, Widget, create_widget
+from pydantic import BaseModel
+from PyQt5.QtWidgets import QScrollArea
 
+default_cell_config = load_config(datasets.cell_config())
 
-def run_tracker(objects, config_file_path):
+# widgets for which the default widget type is incorrect
+hidden_variable_names = [
+    "name",
+    "measurements",
+    "states",
+    "dt",
+    "apoptosis_rate",
+    "prob_not_assign",
+    "eta",
+]
+all_hypotheses = ["P_FP", "P_init", "P_term", "P_link", "P_branch", "P_dead"]
+
+
+@dataclass
+class Matrices:
+    """A helper dataclass to adapt matrix representation to and from pydantic.
+    This is needed because TrackerConfig stored "scaled" matrices, i.e.
+    doesn't store sigma and the "unscaled" matrix separately.
+    """
+
+    names: List[str] = field(default_factory=lambda: ["A", "H", "P", "G", "R"])
+    default_sigmas: List[float] = field(
+        default_factory=lambda: [1.0, 1.0, 150.0, 15.0, 5.0]
+    )
+    unscaled_matrices: Dict[str, List[float]] = field(
+        default_factory=lambda: dict(
+            # stop black from reformatting here
+            # so we can read the matrix entries
+            # fmt: off
+            A_cell=[1, 0, 0, 1, 0, 0,
```

Comment:
It seems really odd that these are stored as lists, and not lists of lists (or 2D arrays) - is there a good implementation reason to do this?

Reply:
I think this is a legacy of the old json config file format, where these arrays needed to be stored flattened. I'm not sure that's an issue now

Reply:
Yea, I implemented this based on the contents of the example JSON config files. I've changed to 2D `numpy` arrays now.